File: source/ch/cyberduck/ui/cocoa/FileController.java
package ch.cyberduck.ui.cocoa;

/*
 *  Copyright (c) 2005 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

import ch.cyberduck.core.Cache;
import ch.cyberduck.core.Path;
import ch.cyberduck.core.Session;
import ch.cyberduck.core.local.BrowserLauncherFactory;
import ch.cyberduck.core.preferences.PreferencesFactory;
import ch.cyberduck.ui.cocoa.application.NSAlert;
import ch.cyberduck.ui.cocoa.application.NSTextField;
import ch.cyberduck.ui.cocoa.foundation.NSRange;

import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.StringUtils;
import org.rococoa.cocoa.foundation.NSRect;
import org.rococoa.cocoa.foundation.NSUInteger;

import java.util.EnumSet;

/**
 * @version $Id$
 */
public abstract class FileController extends AlertController {

    @Outlet
    protected NSTextField inputField
            = NSTextField.textfieldWithFrame(new NSRect(0, 22));

    public void setInputField(final NSTextField inputField) {
        this.inputField = inputField;
    }

    private Cache<Path> cache;

    public FileController(final WindowController parent, final Cache<Path> cache, final NSAlert alert) {
        super(parent, alert);
        this.cache = cache;
        alert.setShowsHelp(true);
    }

    @Override
    public void beginSheet() {
        this.setAccessoryView(inputField);
        super.beginSheet();
    }

    @Override
    protected void focus() {
        this.focus(inputField);
    }

    protected void focus(final NSTextField control) {
        // Focus accessory view.
        control.selectText(null);
        this.window().makeFirstResponder(control);
        control.currentEditor().setSelectedRange(NSRange.NSMakeRange(
                new NSUInteger(0), new NSUInteger(FilenameUtils.getBaseName(control.stringValue()).length())
        ));
    }

    /**
     * @return The current working directory or selected folder
     */
    protected Path getWorkdir() {
        if(((BrowserController) parent).getSelectionCount() == 1) {
            final Path selected = ((BrowserController) parent).getSelectedPath();
            if(null != selected) {
                return selected.getParent();
            }
        }
        return ((BrowserController) parent).workdir();
    }

    protected Path getSelected() {
        return ((BrowserController) parent).getSelectedPath();
    }

    protected Session getSession() {
        return ((BrowserController) parent).getSession();
    }

    @Override
    protected boolean validateInput() {
        if(StringUtils.contains(inputField.stringValue(), Path.DELIMITER)) {
            return false;
        }
        if(StringUtils.isNotBlank(inputField.stringValue())) {
            if(cache.lookup(new Path(this.getWorkdir(), inputField.stringValue(), EnumSet.of(Path.Type.file)).getReference()) != null) {
                return false;
            }
            if(cache.lookup(new Path(this.getWorkdir(), inputField.stringValue(), EnumSet.of(Path.Type.directory)).getReference()) != null) {
                return false;
            }
            return true;
        }
        return false;
    }

    @Override
    protected void help() {
        final StringBuilder site = new StringBuilder(PreferencesFactory.get().getProperty("website.help"));
        site.append("/howto/browser");
        BrowserLauncherFactory.get().open(site.toString());
    }
}

File: windows/src/main/csharp/ch/cyberduck/ui/controller/FileController.cs
// 
// Copyright (c) 2010-2016 Yves Langisch. All rights reserved.
// http://cyberduck.io/
// 
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// Bug fixes, suggestions and comments should be sent to:
// feedback@cyberduck.io
// 

using System.Drawing;
using System.Text.RegularExpressions;
using System.Windows.Forms;
using ch.cyberduck.core;
using ch.cyberduck.core.pool;
using ch.cyberduck.core.preferences;
using Ch.Cyberduck.Core;
using java.util;
using static Ch.Cyberduck.ImageHelper;

namespace Ch.Cyberduck.Ui.Controller
{
    internal abstract class FileController : WindowController<IPromptView>
    {
        protected readonly BrowserController BrowserController;
        private string _input = string.Empty;

        protected FileController(IPromptView view, BrowserController browserController)
        {
            BrowserController = browserController;
            View = view;

            View.ValidateInput += ValidateInput;
        }

        public virtual Image IconView
        {
            get { return IconProvider.GetFileIcon("_unknown", false, true, false); }
        }

        public virtual Path Workdir
        {
            get
            {
                if (BrowserController.SelectedPaths.Count == 1)
                {
                    return BrowserController.SelectedPath.getParent();
                }
                return BrowserController.Workdir;
            }
        }

        protected string PrefilledText
        {
            get { return _input; }
            set { _input = value; }
        }

        protected SessionPool Session
        {
            get { return BrowserController.Session; }
        }

        protected virtual bool ValidateInput()
        {
            string t = View.InputText.Trim();
            foreach (string f in Utils.ConvertFromJavaList<string>(PreferencesFactory.get().getList("browser.filter.regex")))
            {
                if (Regex.IsMatch(t, f))
                {
                    return false;
                }
            }
            if (!string.IsNullOrEmpty(t))
            {
                if (
                    BrowserController.Cache.get(BrowserController.Workdir)
                        .contains(new Path(BrowserController.Workdir, t, EnumSet.of(AbstractPath.Type.file))))
                {
                    return false;
                }
                if (
                    BrowserController.Cache.get(BrowserController.Workdir)
                        .contains(new Path(BrowserController.Workdir, t, EnumSet.of(AbstractPath.Type.directory))))
                {
                    return false;
                }
                return true;
            }
            return false;
        }

        public abstract void Callback(DialogResult result);

        public void Show()
        {
            View.InputText = PrefilledText;
            View.IconView = IconView;
            Callback(View.ShowDialog(BrowserController.View));
        }
    }
}
